pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1152 @@
+		// cannot be resolved by triggering a bot
+		ensure!(
+			RedemptionPayoutCurrency::<T>::get(who, investment_id)
+				.map(|currency| currency == payout_currency)
+				.unwrap_or_else(|| {
+					log::debug!("Redemption payout currency missing when calling decrease. Should never occur if redemption has been increased beforehand");
+					false
+				}),
+			Error::<T>::InvalidRedemptionPayoutCurrency
+		);
+		// TODO(future): Add implicit collection or error handling (i.e. message to
+		// source domain)
+		ensure!(
+			!T::Investment::redemption_requires_collect(who, investment_id),
+			Error::<T>::RedeemError(RedeemError::CollectRequired)
+		);
+
+		let pre_state = RedemptionState::<T>::get(who, investment_id);
+		let post_state = pre_state
+			.transition(RedeemTransition::DecreaseRedeemOrder(amount))
+			.map_err(|e| {
+				log::debug!("RedeemState transition error: {:?}", e);
+				Error::<T>::from(RedeemError::Decrease)
+			})?;
+		Pallet::<T>::apply_redeem_state_transition(who, investment_id, post_state)?;
+
+		Ok(amount)
+	}
+
+	#[transactional]
+	fn collect_foreign_investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		foreign_payout_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<ExecutedForeignCollectInvest<T::Balance>, DispatchError> {
+		// Note: We assume the configured Investment trait to notify about the collected
+		// amounts via the `CollectedInvestmentHook` which handles incrementing the
+		// `CollectedInvestment` amount.
+		T::Investment::collect_investment(who.clone(), investment_id)?;
+
+		Self::transfer_collected_investment(
+			who,
+			investment_id,
+			foreign_payout_currency,
+			pool_currency,
+		)
+	}
+
+	#[transactional]
+	fn collect_foreign_redemption(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		foreign_payout_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<(), DispatchError> {
+		ensure!(
+			RedemptionPayoutCurrency::<T>::get(who, investment_id)
+				.map(|currency| currency == foreign_payout_currency)
+				.unwrap_or_else(|| {
+					log::debug!("Corruption: Redemption payout currency missing when calling decrease. Should never occur if redemption has been increased beforehand");
+					false
+				}),
+			Error::<T>::InvalidRedemptionPayoutCurrency
+		);
+		ensure!(T::PoolInspect::currency_for(investment_id.of_pool())
+			.map(|currency| currency == pool_currency)
+			.unwrap_or_else(|| {
+				log::debug!("Corruption: Failed to derive pool currency from investment id when collecting foreign redemption. Should never occur if redemption has been increased beforehand");
+				false
+			}),
+			DispatchError::Corruption
+		);
+
+		// Note: We assume the configured Investment trait to notify about the collected
+		// amounts via the `CollectedRedemptionHook` which handles incrementing the
+		// `CollectedRedemption` amount.
+		T::Investment::collect_redemption(who.clone(), investment_id)?;
+
+		Ok(())
```

Comment:
I would expect here an analogous `transfer_collected_redeemption()` (as found in `collect_foreign_investment`) where to transfer funds to the investor, with the `stable_to_stable()` call inside of it. Why is not needed in this case?

Reply:
The difference between collecting an investment and a redemption is that the former is synchronous as it does not involve any swapping, it simply collects tranche tokens and transfers them to the domain locator, the receiving domain takes care of accounting the tranche tokens representation.

However, the latter is asynchronous, see [flow in V3 spec](https://centrifuge.hackmd.io/@Frederik/S1fmJty_s#Increase-amp-Collect-Redeem-Order-from-DAI-investment-of-USDC-Pool). After collecting a non-zero redemption amount, it needs to be swapped from pool currency to foreign currency. Only once that swap is fulfilled (i.e. the `InnerRedeemState` includes `SwapIntoForeignDone` without `ActiveSwapIntoForeignCurrency`), the `ExecutedCollectRedeem` message can be sent to the other domain. This burns the `foreign` currency on the Centrifuge Chain side and transfers the amount on the recipient domain.

Reply:
Thanks for the long explanation